File: ios/Sources/CapacitorLlamaPlugin/CapacitorLlamaPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CapacitorLlamaPlugin)
public class CapacitorLlamaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorLlamaPlugin"
    public let jsName = "CapacitorLlama"
    // TODO: implement the bridge pattern
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initContext", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseContext", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopCompletion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseAllContexts", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getFormattedChat", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "completion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "tokenize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detokenize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVocab", returnType: CAPPluginReturnPromise)
    ]

    @objc func initContext(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            let result = CAPLlama.initContext(contextId, withContextParams: call.options)
            
            if let resultDict = result as? [String: Any] {
                call.resolve(resultDict)
            } else {
                call.reject("Invalid result format from CapacitorLlamaPlugin.initContext")
            }

        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func releaseContext(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            CAPLlama.releaseContext(contextId)
            call.resolve()
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }
    
    @objc func stopCompletion(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            CAPLlama.stopCompletion(contextId)
            call.resolve()
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func releaseAllContexts(_ call: CAPPluginCall) {
        do {
            CAPLlama.releaseAllContexts()
            call.resolve()
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func getFormattedChat(_ call: CAPPluginCall) {
        let messages = call.getString("messages")
        let chatTemplate = call.getString("chatTemplate")
        guard let contextId = call.getDouble("id") else {
            call.reject("Missing required parameter 'id'.")
            return
        }
        do {
            let result = CAPLlama.getFormattedChat(contextId, withMessages: messages, withTemplate: chatTemplate, withParams: call.options)
            if let resultDict = result as? [String: Any] {
                call.resolve(resultDict)
            } else {
                call.reject("Invalid result format from CapacitorLlamaPlugin.initContext")
            }
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func completion(_ call: CAPPluginCall) {
        guard let contextId = call.getDouble("id") else {
            call.reject("Missing required parameter 'id'.")
            return
        }
        do {
            var params = call.getObject("params");
            // Safely get the "messages" array
            let messages = params?["messages"] as? [[String: Any]]
            if (messages != nil && params != nil) {
                // Convert the messages array into JSON data
                let jsonData = try JSONSerialization.data(withJSONObject: messages!, options: [])
                // Convert JSON data to a String (if you need a JSON string)
                let jsonString = String(data: jsonData, encoding: .utf8)

                print("Messages JSON String: \(jsonString ?? "nil")")
                    
                // Now you have `jsonData` or `jsonString` to work with
                
                let chatTemplate = params?["chatTemplate"] as? String
                let formatResult = CAPLlama.getFormattedChat(contextId, withMessages: jsonString, withTemplate: chatTemplate, withParams: params)
                params!["prompt"] = formatResult?["prompt"] as! String as String
            }

          CAPLlama.completion(contextId, withCompletionParams: params, onToken: { (tokenData: NSMutableDictionary?) in
                guard let tokenData = tokenData else { return }
                self.notifyListeners("onToken", data: ["contextId": contextId, "tokenResult": tokenData])
            }, call: call)     
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func tokenize(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            guard let text = call.getString("text") else {
                call.reject("Missing required parameter 'text'.")
                return
            }
            let result = CAPLlama.tokenize(contextId, text: text)
             if let resultDict = result as? [String: Any] {
                call.resolve(resultDict)
            } else {
                call.reject("Invalid result format from CapacitorLlamaPlugin.tokenize")
            }
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func detokenize(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            guard let tokens = call.getArray("tokens") else {
                call.reject("Missing required parameter 'tokens'.")
                return
            }
            let result = CAPLlama.detokenize(contextId, tokens: tokens)
             if let resultDict = result as? [String: Any] {
                call.resolve(resultDict)
            } else {
                call.reject("Invalid result format from CapacitorLlamaPlugin.detokenize")
            }
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }

    @objc func getVocab(_ call: CAPPluginCall) {
        do {
            guard let contextId = call.getDouble("id") else {
                call.reject("Missing required parameter 'contextId'.")
                return
            }
            let result = CAPLlama.getVocab(contextId)
             if let resultDict = result as? [String: Any] {
                call.resolve(resultDict)
            } else {
                call.reject("Invalid result format from CapacitorLlamaPlugin.tokenize")
            }
        } catch let error as NSException {
            call.reject(error.reason ?? "Unknown error")
        }
    }
}

File: android/src/main/java/com/capllama/CapacitorLlamaPlugin.java
package com.capllama;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.concurrent.CompletableFuture;

@CapacitorPlugin(name = "CapacitorLlama")
public class CapacitorLlamaPlugin extends Plugin {

    private CapacitorLlama implementation = new CapacitorLlama();

    @PluginMethod
    public void initContext(PluginCall call) {
        Integer id = call.getInt("id", -1);
        CompletableFuture<JSObject> future = implementation.initContext(id, call.getData());
        future.thenAccept(call::resolve);
    }

    private class PartialCompletionCallbackWithListeners implements PartialCompletionCallback {

        int contextId;
        boolean emitNeeded;

        public PartialCompletionCallbackWithListeners(int contextId, boolean emitNeeded) {
            this.contextId = contextId;
            this.emitNeeded = emitNeeded;
        }

        public void onPartialCompletion(JSObject tokenResult) {
            if (this.emitNeeded) {
                JSObject ret = new JSObject();
                ret.put("contextId", this.contextId);
                ret.put("tokenResult", tokenResult);
                notifyListeners("onToken", ret);
            }
        }
    }

    @PluginMethod
    public void completion(PluginCall call) {
        JSObject params = call.getData();
        JSObject innerParams = params.getJSObject("params");
        CompletableFuture<JSObject> future = implementation.completion(
            params,
            new PartialCompletionCallbackWithListeners(params.optInt("id", 0), innerParams.optBoolean("emit_partial_completion", false))
        );
        future.thenAccept(call::resolve);
    }

    @PluginMethod
    public void getFormattedChat(PluginCall call) {
        CompletableFuture<JSObject> future = implementation.getFormattedChat(call.getData());
        future.thenAccept(call::resolve);
    }

    @PluginMethod
    public void releaseAllContexts(PluginCall call) {
        CompletableFuture<Void> future = implementation.releaseAllContexts();
        future.thenAccept((v) -> call.resolve());
    }

    @PluginMethod
    public void releaseContext(PluginCall call) {
        Integer id = call.getInt("id", -1);
        CompletableFuture<Void> future = implementation.releaseContext(id, call.getData());
        future.thenAccept((v) -> call.resolve());
    }

    @PluginMethod
    public void stopCompletion(PluginCall call) {
        Integer id = call.getInt("id", -1);
        CompletableFuture<Void> future = implementation.stopCompletion(id);
        future.thenAccept((v) -> call.resolve());
    }

    @PluginMethod
    public void detokenize(PluginCall call) {
        CompletableFuture<JSObject> future = implementation.detokenize(call.getData());
        future.thenAccept(call::resolve);
    }

    @PluginMethod
    public void tokenize(PluginCall call) {
        CompletableFuture<JSObject> future = implementation.tokenize(call.getData());
        future.thenAccept(call::resolve);
    }

    @PluginMethod
    public void getVocab(PluginCall call) {
        CompletableFuture<JSObject> future = implementation.getVocab(call.getData());
        future.thenAccept(call::resolve);
    }
}
